crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -185,56 +238,1250 @@ impl<S: States> NextState<S> {
+///     fn compute(sources: Option<AppState>) -> Option<Self> {
+///         match sources {
+///             /// When we are in game, we want to return the InGame state
+///             Some(AppState::InGame { .. }) => Some(InGame),
+///             /// Otherwise, we don't want the `State<InGame>` resource to exist,
+///             /// so we return None.
+///             _ => None
+///         }
+///     }
+/// }
+///
+/// #[derive(SubStates, Clone, PartialEq, Eq, Hash, Debug, Default)]
+/// #[source(InGame = InGame)]
+/// enum GamePhase {
+///     #[default]
+///     Setup,
+///     Battle,
+///     Conclusion
+/// }
+/// ```
+///
+/// However, you can also manually implement them. Note that you'll also need to manually implement the `States` & `FreelyMutableState` traits.
+/// Unlike the derive, this does not require an implementation of [`Default`], since you are providing the `exists` function
+/// directly.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # use bevy_ecs::schedule::FreelyMutableState;
+///
+/// /// Computed States require some state to derive from
+/// #[derive(States, Clone, PartialEq, Eq, Hash, Debug, Default)]
+/// enum AppState {
+///     #[default]
+///     Menu,
+///     InGame { paused: bool }
+/// }
+///
+/// #[derive(Clone, PartialEq, Eq, Hash, Debug)]
+/// enum GamePhase {
+///     Setup,
+///     Battle,
+///     Conclusion
+/// }
+///
+/// impl SubStates for GamePhase {
+///     /// We set the source state to be the state, or set of states,
+///     /// we want to depend on. Any of the states can be wrapped in an Option.
+///     type SourceStates = Option<AppState>;
+///
+///     /// We then define the compute function, which takes in the [`Self::SourceStates`]
+///     fn exists(sources: Option<AppState>) -> Option<Self> {
+///         match sources {
+///             /// When we are in game, so we want a GamePhase state to exist, and the default is
+///             /// GamePhase::Setup
+///             Some(AppState::InGame { .. }) => Some(GamePhase::Setup),
+///             /// Otherwise, we don't want the `State<GamePhase>` resource to exist,
+///             /// so we return None.
+///             _ => None
+///         }
+///     }
+/// }
+///
+/// impl States for GamePhase {
+///     const DEPENDENCY_DEPTH : usize = <GamePhase as SubStates>::SourceStates::SET_DEPENDENCY_DEPTH + 1;
+/// }
+///
+/// impl FreelyMutableState for GamePhase {}
+/// ```
+pub trait SubStates: States + FreelyMutableState {
+    /// The set of states from which the [`Self`] is derived.
+    ///
+    /// This can either be a single type that implements [`States`], or a tuple
+    /// containing multiple types that implement [`States`], or any combination of
+    /// types implementing [`States`] and Options of types implementing [`States`]
+    type SourceStates: StateSet;
+
+    /// This function gets called whenever one of the [`SourceStates`](Self::SourceStates) changes.
```

Comment:
The structure of this paragraph is backwards. This should declare what the method does first, and later describe when it's called.

Reply:
changed!